### PR TITLE
[ACCUMULO-4147] - Add Accumulo Namespace Operations to Proxy

### PR DESCRIPTION
**Pull request by tristeng**
I've updated the the Accumulo Proxy to support namespace operations (including the security operations). I have tested all the new functions using the Python bindings on a mini cluster.

### CODE REVIEW

**Comment by keith-turner** [comment]
@tristeng  the proxy has fairly comprehensive ITs that excercise most of its API.  There are located in `accumulo-test/src/main/java/org/apache/accumulo/test/proxy`.  Would you be able to add test for these new namespace operations?

**Comment by tristeng** [comment]
Whoops! Didn't even see those...was only looking under the accumulo-proxy project. Yup, will do. I'll also add in my Python IT file into the proxy/examples/python directory.

**Comment by tristeng** [comment]
Added in IT tests for the new functions, but ran into an interesting error and am hoping you all can provide me with some insight. I've added the following test into TestProxyNamespaceOperations:
```
  @Ignore("namespaceIdMap() is returning unexpected results; Python client works as expected. Investigating")
  @Test
  public void namespaceIds() throws TException {
    assertTrue(tpc.proxy().namespaceIdMap(userpass).containsKey("accumulo"));
    assertEquals(tpc.proxy().namespaceIdMap(userpass).get("accumulo"), "+accumulo");
  }
```


You can see I've ignored it...when I run through the debugger, `tpc.proxy().namespaceIdMap(userpass)` returns a map that seems to list the accumulo namespace tables (unexpected):
```
"accumulo.replication" -> "accumulo.replication"
"accumulo.metadata" -> "accumulo.metadata"
"accumulo.root" -> "accumulo.root"
```

When I run this same function through the Python test client and print the map, I get this Python map (which is what I expect for this function):
`{'': '+default', 'testing': '1', 'accumulo': '+accumulo'}`

As far as I can tell, I haven't messed up the function in the ProxyServer.java file:
```
  @Override
  public Map<String,String> namespaceIdMap(ByteBuffer login) throws org.apache.accumulo.proxy.thrift.AccumuloException,
      org.apache.accumulo.proxy.thrift.AccumuloSecurityException, TException {
    try {
      return getConnector(login).namespaceOperations().namespaceIdMap();
    } catch (Exception e) {
      handleException(e);
      return null;
    }
  }
```

Any thoughts on where I should look?

**Comment by keith-turner** [comment]
> Any thoughts on where I should look?

Looking at the test I have a strong suspicion that the reason the test is failing is because its using MockAccumulo.   When things worked w/ your Python test client, was that proxy back a mini or real accumulo instance?

I would discourage using Mock.  Its been deperecated because it was not being well maintained as features (like namspaces) were added to Accumulo.  I would also encourage you to add your namespace test to SimpleProxyBase.  TBinaryProxyIT, TJsonProtocolProxyIT, etc all extend SimpleProxyBase but run configure a different protocol.  All of the test in SimpleProxyBase are run using multiple thrift protocols.   Also, the test in SimpleProxyBase use MiniAccumulo instead of MockAccumulo.

**Comment by tristeng** [comment]
Yes, my Python integration tests were run on a mini Accumulo cluster that I had started myself and judging by how fast the tests run in the new TestProxyNamespaceOperations class I would have to assume its using the mock instance. I'll add my tests to SimpleProxyBase, thanks!

**Comment by tristeng** [comment]
I've added in integration tests and everything is passing. Let me know if you think I should add anything else to finish close this issue.

File: test/src/main/java/org/apache/accumulo/test/proxy/SimpleProxyBase.java [hunk 1]
```diff
@@ -818,6 +950,58 @@ public void tableNotFound() throws Exception {
   }
 
   @Test
+  public void namespaceNotFound() throws Exception {
+    final String doesNotExist = "doesNotExists";
+    try {
+      client.deleteNamespace(creds, doesNotExist);
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/proxy/SimpleProxyBase.java]: If this thrift call does not throw a NNFE, it seems like the test will still pass?  If this is the case, could place an Assert.fail() after each thrift call thats supposed to throw.

File: test/src/main/java/org/apache/accumulo/test/proxy/SimpleProxyBase.java [hunk 2]
```diff
@@ -836,6 +1020,22 @@ public void testExists() throws Exception {
   }
 
   @Test
+  public void testNamespaceExists() throws Exception {
+    try {
+      client.createNamespace(creds, namespaceName);
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/proxy/SimpleProxyBase.java]: Also, will this test pass if not exception is thrown?

**Comment by keith-turner** [comment]
The changes look great.  I am finished looking at this and have nothing else to add.

**Comment by joshelser** [comment]
@keith-turner do you have time to merge this in? I can try to swing around on it soon if not.

**Comment by tristeng** [comment]
I'm about to add in the 2 changes Keith suggested...should only take me a few minutes, and then however long it takes to run the tests.

**Comment by keith-turner** [comment]
Thanks @tristeng.  I squashed and merge this, can you close this PR?

I ran the following test w/ no problem before pushing.  That caused some reformatting which was included in the commit I pushed.  I also added `since 1.8.0` to the comment in the proxy.thrift file.

```
mvn clean verify -DskipITs
mvn -Dit.test=".*proxy.*" -Dtest=foo -Dfindbugs.skip -Dcheckstyle.skip -DfailIfNoTests=false verify
```

**Comment by tristeng** [comment]
Thanks!

**Pull request closed without merging**